### ISSUE TITLE
[Bug]: https://api.nuget.org/v3/index.json/ returning 404 (with trailing forward slash)

### ISSUE THREAD

**Issue by ac-proagrica** [NuGet/NuGetGallery#9227]
### NuGet Product Used

Other/NA

### Product Version

Nuget Service Index

### Worked before?

Yes

### Repro Steps & Context

Nexus adds a trailing slash to the service index url and expects both https://api.nuget.org/v3/index.json/ and https://api.nuget.org/v3/index.json to return the same response.

This can be seen in the answer to a similar issue here: https://issues.sonatype.org/browse/NEXUS-25357

The URL with the trailing backslash is not working today and we're unable to connect.

### Verbose Logs

```shell
Nexus logs:
2022-08-26 11:08:27,231+0000 DEBUG [qtp897030751-965] admin com.sonatype.nexus.repository.nuget.orient.internal.OrientNugetProxyFacet - Fetching: GET https://api.nuget.org/v3/index.json/ HTTP/1.1
2022-08-26 11:08:27,232+0000 DEBUG [qtp897030751-965] admin org.sonatype.nexus.httpclient.outbound - https://api.nuget.org/v3/index.json/ > GET /v3/index.json/ HTTP/1.1
2022-08-26 11:08:27,232+0000 DEBUG [qtp897030751-965] admin org.sonatype.nexus.internal.httpclient.SharedHttpClientConnectionManager - Connection request: [route: {s}->https://api.nuget.org:443][total available: 2; route allocated: 0 of 20; total allocated: 2 of 200]
2022-08-26 11:08:27,232+0000 DEBUG [qtp897030751-965] admin org.sonatype.nexus.internal.httpclient.SharedHttpClientConnectionManager - Connection leased: [id: 638][route: {s}->https://api.nuget.org:443][total available: 2; route allocated: 1 of 20; total allocated: 3 of 200]
2022-08-26 11:08:27,409+0000 DEBUG [qtp897030751-965] admin org.sonatype.nexus.httpclient.outbound - https://api.nuget.org/v3/index.json/ < HTTP/1.1 404 Not Found @ 177.7 ms
2022-08-26 11:08:27,410+0000 DEBUG [qtp897030751-965] admin com.sonatype.nexus.repository.nuget.orient.internal.OrientNugetProxyFacet - Response: HttpResponseProxy{HTTP/1.1 404
```

**Comment by joelverhagen** [comment]
Thanks for the heads up @donnie-msft. I transferred it.

**Comment by joelverhagen** [comment]
Hello @ac-proagrica, I've read through your issue. 

Given our current V3 protocol, it is acceptable for the NuGet.org feed to return a 404 Not Found for the URL `https://api.nuget.org/v3/index.json/`. This is not the documented service index and there is no NuGet V3 protocol documentation that asserts the slash and unslashed version of the URL should be treated as the same by the server.

_That being said,_ NuGet.org currently runs several CDN providers each with slightly different rules engines. There is currently mixed behavior on this specific case which results in some users (depending on customer geo and failover conditions) to see this URL returning a 200 OK. Azure China CDN and Verizon CDN allow this pattern. Microsoft CDN does not.

It is not ideal for NuGet.org to have inconsistent behavior for this URL but it is also not breaking the protocol. Essentially this is behavior outside of the specification and is therefore undefined (i.e. not guaranteed to work a certain way).

From an intuitive perspective, you can look at the forward slashes in the path part of the URL as directory separators. The variant of the URL you provide that ends in a forward slash is implying that "index.json/" is a directory not a file. It is not a directory in our mental model or in actuality (this is a blob in Azure Blob storage with the file name "index.json"). Note that this is just an analogy. Comparing a file system to URLs is fraught with problems and will probably lead to more trouble than good in the end, but I wanted to provide this perspective for the sake of understanding.

I would assert that the behavior from Nexus is incorrect. I do not have context on why they made that decision, but I am going to guess that the right thing to happen here is for Nexus to at least disable this behavior for host names ending in ".nuget.org" or perhaps for all feeds.

It seems like _any_ munging/manipulation of the source URL provided to the proxy will cause problems. You could maybe make an argument for swapping HTTP for HTTPS for security reasons but assumptions like this generally should not be made by a NuGet client (the Nexus proxy implementation in this case).

From a strict protocol sense, the V3 source URL should be used exactly like how the [package source maintainer documents](https://docs.microsoft.com/en-us/nuget/api/overview#service-index): https://api.nuget.org/v3/index.json. Anything outside of this is undocumented behavior and cannot be guaranteed from one source implementation to the next.

I should note that there are other variants of the NuGet.org V3 service index URL different from the documented one that may work now or did at some point in the past. _But this is not behavior that should be relied upon and is likely due to some artifact or limitation of our current implementation._

As the NuGet.org team changes underlying architecture, we focus on the API contract between NuGet client and NuGet server. The API contract _really_ is whatever the official NuGet Client software expects and is (to our best knowledge) documented here: https://docs.microsoft.com/en-us/nuget/api/overview. We certainly consider other client software as well as it is our job to be good stewards of the ecosystem, but I don't have the data to show that this is a quirk that we should officially support. To be honest, this is the first I heard of this behavior after nearly 7 years on the team.

Attempting to reimplement all bugs/oddities that we have ever shipped but are not depended upon the clients in our primary test matrix (official NuGet client software) is not sustainable for the team.

```powershell
PS C:\Users\jver\Downloads> .\nuget.exe search json -source "https://api.nuget.org/v3/index.json/" -take 1 -verbosity normal
  GET https://api.nuget.org/v3/index.json/Search()?$filter=IsLatestVersion&searchTerm='json'&targetFramework=''&includePrerelease=false&$skip=0&$top=1&semVerLevel=2.0.0
  OK https://api.nuget.org/v3/index.json/Search()?$filter=IsLatestVersion&searchTerm='json'&targetFramework=''&includePrerelease=false&$skip=0&$top=1&semVerLevel=2.0.0 30ms
Data at the root level is invalid. Line 1, position 1.
PS C:\Users\jver\Downloads> .\nuget.exe search json -source "https://api.nuget.org/v3/index.json" -take 1 -verbosity normal
====================
Source: nuget.org
--------------------
> Newtonsoft.Json | 13.0.1 | Downloads: 2,223,138,823
  Json.NET is a popular high-performance JSON framework for .NET
--------------------
```

As you can see, official NuGet client software does not have this behavior.

All this being said, I'm happy to leave this issue open and continue the conversation with you or the folks that maintain Nexus. It might be possible for us to support for this non-standard URL but we would have to assess feasibility with our CDN providers and we'd also need to make sure this was the right use of the team's time given our other priorities.

**Comment by ac-proagrica** [comment]
Hello @joelverhagen
Thank you for taking the time to provide such a detailed response. It's very informative. 

It's very difficult to argue with what you've outlined, I have just not been able to find the basis for the assumption that has been made by the Nexus team, nor been able to understand why it suddenly "broke" but it may be related to what you've said about CDN providers.

I have also found a way to fix this. I think what's happened here is a combination of a bug in the Nexus version that we use, which we've only noticed because we were (possibly) getting a 200 on the wrong endpoint. 

I'm happy for this issue to be closed.
